File: pygame/game.py
#!/usr/bin/env python3

import pygame
from sys import exit

from settings import *
from background import Background
from player import Player
from base import Base
from pipe_spawner import PipeSpawner

from message_screen import MessageScreen
from game_over_screen import GameOverScreen

class Game:
    # =========================================================
    def __init__(self):
        self.clock = None
        self.window = None
        self.background = None
        self.is_started = False
        self.is_died = False
        self.score = 0
        self.high_score = 0
        
    # =========================================================
    def init(self) -> None:
        pygame.init()
        self.font = pygame.font.Font(RESOURSE_FONT_PATH, 45)

        self.clock = pygame.time.Clock()
        self.window = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT))

        pygame.display.set_caption(WINDOW_CAPTION)

        self.message_screen:MessageScreen = MessageScreen()
        self.game_over_screen:GameOverScreen = GameOverScreen(self.font)

        self.on_restart()

    # =========================================================
    def main(self) -> None:
        running:bool = True
        while running:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    running = False
                    pygame.quit()
                    exit()
                elif event.type == pygame.MOUSEBUTTONDOWN:
                    if self.is_died == True:
                        if self.game_over_screen.is_restart_btn_clicked(event):
                            self.on_restart()
                    else:
                        if event.button == 1:
                            # 1 = Left, 2 = Middle, 3 = Right, 4 = Scroll Up, 5 = Scroll Down
                            self.on_flap_action()
                elif event.type == pygame.KEYDOWN:
                    # Check if that specific key was the Spacebar
                    if event.key == pygame.K_SPACE:
                        self.on_flap_action()
            
            self.update(self.window)
            self.draw(self.window)
            pygame.display.update()
            self.clock.tick(WINDOW_FPS)

    # =========================================================
    def update(self, window) -> None:
        self.background.update(window)
        self.base.update(window)
        self.pipe_spawner.update(window)
        self.player_group.update(window)

        # Collision detection
        collision_pipes = pygame.sprite.spritecollide(self.player_group.sprites()[0], self.pipe_spawner.pipes, False)
        collision_ground = self.player_group.sprites()[0].rect.colliderect(self.base.rect)
        collision = collision_pipes or collision_ground
        if collision:
            self.on_game_over()

    # =========================================================
    def draw(self, window) -> None:
        self.background.draw(window)
        self.pipe_spawner.draw(window)
        self.base.draw(window)
        self.player_group.draw(window)

        if self.is_started == True:
            # Draw score text
            score_text = self.font.render(str(self.score), True, pygame.Color(255, 255, 255))
            window.blit(score_text, (20, 20))
        else:
            if self.is_died == False:
                self.message_screen.draw(window)
            else:
                self.game_over_screen.draw(window)
                

    # =========================================================
    def on_flap_action(self):
        if self.is_died == False:
            if self.is_started == False:
                self.is_started = True
                self.on_start()
            self.on_flap()

    # =========================================================
    def on_flap(self):
        self.player.on_flap()

    # =========================================================
    def on_start(self):
        self.score = 0
        self.base.on_start()
        self.pipe_spawner.on_start()
        self.player.on_start()

    # =========================================================
    def on_restart(self):
        self.background:Background = Background()
        
        self.base:Base = Base()

        self.player:Player = Player()
        self.player_group = pygame.sprite.GroupSingle()
        self.player_group.add(self.player)

        self.pipe_spawner:PipeSpawner = PipeSpawner()
        self.pipe_spawner.score_listener = self.on_score

        self.is_died = False
        self.is_started = False

    # =========================================================
    def on_game_over(self):
        self.is_died = True
        self.is_started = False
        if self.score > self.high_score:
            self.high_score = self.score
        self.game_over_screen.set_score(self.score, self.high_score)
        self.base.on_game_over()
        self.pipe_spawner.on_game_over()
        self.player.on_game_over()

    # =========================================================
    def on_score(self):
        # print("Score!")
        self.score += 1

File: sfml/src/Game.cpp
#include <iostream>
#include "Game.h"

#include "settings.h"

Game :: Game() : 
    window(sf::VideoMode(WINDOW_WIDTH, WINDOW_HEIGHT), WINDOW_CAPTION, sf::Style::Titlebar | sf::Style::Close),
    background(),
    base(),
    player(),
    pipe_spawner(),
    messageScreen(),
    gameOverScreen(),
    font(), scoreText(),
    is_started(false), is_died(false),
    score(0), high_score(0)
{
    window.setFramerateLimit(WINDOW_FPS);
    pipe_spawner.scoreListener = std::bind(&Game::onScore, this);

    if(!font.loadFromFile(RESOURCE_FONT_PATH))
    {
         throw std::runtime_error(
            "Failed to load game font");
    }

    scoreText.setFont(font);
    scoreText.setCharacterSize(45);
    scoreText.setFillColor(sf::Color::White);
    scoreText.setPosition(20.f, 20.f);
}

void Game :: init()
{
}

int Game :: main(){
    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            switch(event.type)
            {
                case sf::Event::Closed:
                {
                    window.close();
                    break;
                }
                case sf::Event::Resized:
                {
                    sf::FloatRect visibleArea(
                    0.f,
                    0.f,
                    static_cast<float>(event.size.width),
                    static_cast<float>(event.size.height));

                    window.setView(sf::View(visibleArea));
                    break;
                }
                case sf::Event::MouseButtonPressed:
                {
                    if (event.mouseButton.button == sf::Mouse::Left)
                    {
                        if(is_died)
                        {
                            if(gameOverScreen.isRestartClicked(event, window))
                                onRestart();
                        }
                        else
                        {
                            onFlapAction();
                        }
                    }
                    break;
                }
                case sf::Event::KeyPressed:
                {
                    if (event.key.code == sf::Keyboard::Space)
                    {
                        onFlapAction();
                    }
                    break;
                }
                default:
                {
                    break;
                }
            }
        }

        this->update();

        window.clear();

        this->draw();

        window.display();
    }
    return 0;
}

void Game :: update()
{
    background.update(window);
    base.update(window);
    pipe_spawner.update();
    player.update(window);

    bool collision = collisionDetect();
    if(collision)
        onGameOver();
}

void Game :: draw()
{
    background.draw(window);
    pipe_spawner.draw(window);
    base.draw(window);
    player.draw(window);

    if(is_started)
    {
        scoreText.setString(std::to_string(score));
        window.draw(scoreText);
    }
    else
    {
        if(is_died == false)
        {
            messageScreen.draw(window);
        }
        else
        {
            gameOverScreen.setScore(score, high_score);
            gameOverScreen.draw(window);
        }
    }
}

bool Game :: collisionDetect() const
{
    sf::FloatRect playerRect = player.getBounds();
    sf::FloatRect baseRect = base.getBounds();
    bool collideWithGround = baseRect.intersects(playerRect);
    if(collideWithGround)
        return true;

    bool collideWithPipe = pipe_spawner.pipeCollideWithRect(playerRect);
    return collideWithPipe;
}

void Game :: onFlapAction()
{
    if(is_died == false)
    {
        if(is_started == false)
        {
            is_started = true;
            onStart();
        }
        onFlap();
    }
}

void Game :: onFlap()
{
    player.onFlap();
}

void Game :: onStart()
{
    score = 0;
    base.onStart();
    pipe_spawner.onStart();
    player.onStart();
}

void Game :: onRestart()
{
    base.onRestart();
    player.onRestart();
    pipe_spawner.onRestart();
    is_died = false;
    is_started = false;
}

void Game :: onGameOver()
{
    is_died = true;
    is_started = false;
    if(score > high_score)
        high_score = score;
    // game_over_screen.set_score(score, high_score)
    base.onGameOver();
    pipe_spawner.onGameOver();
    player.onGameOver();
}

void Game :: onScore()
{
    score++;
    std :: cout << "Soce: " << score << std :: endl;
}